### PR TITLE
Fix issue #8823: evil-paste-pop doesn't work `previous command was not an evil paste.`

### PR DESCRIPTION
**Pull request by nochiel**
Fix issue #8823.

[evil-common.el: evil-paste-pop](https://github.com/emacs-evil/evil/blob/master/evil-common.el#L2604), has the following check:

```
(unless (memq last-command
                '(evil-paste-after
                  evil-paste-before
                  evil-visual-paste))
    (user-error "Previous command was not an evil-paste: %s" last-command))
```
Calls to [spacemacs/evil-mc-paste-after](https://github.com/syl20bnr/spacemacs/blob/develop/layers/%2Bspacemacs/spacemacs-evil/funcs.el#L87) and [spacemacs/evil-mc-paste-before](https://github.com/syl20bnr/spacemacs/blob/develop/layers/%2Bspacemacs/spacemacs-evil/funcs.el#L94) thus break `evil-paste-pop`.  A more idiomatic way to create this functionality is to use [elisp advice](https://www.gnu.org/software/emacs/manual/html_node/elisp/Advising-Functions.html).

### CODE REVIEW

File: layers/+spacemacs/spacemacs-evil/funcs.el [hunk 1]
```diff
@@ -84,16 +84,20 @@ Otherwise, revert to the default behavior (i.e. enable `evil-insert-state')."
...
            (eq (evil-mc-get-cursor-count) 1))))
 
-(defun spacemacs/evil-mc-paste-after (&optional count register)
-  "Disable paste transient state if there is more that 1 cursor."
-  (interactive "p")
-  (if (spacemacs//paste-transient-state-p)
-      (spacemacs/paste-transient-state/evil-paste-after)
-    (evil-paste-after count (or register evil-this-register))))
+(defun spacemacs/evil-mc-paste-after (fn &rest args)
+   "Disable paste transient state if there is more that 1 cursor."
```
Comment (duianto) [inline layers/+spacemacs/spacemacs-evil/funcs.el]: The extra space before these lines should probably be removed. Two spaces seem to be default. At least when one auto indents the paragraph with `evil-indent`: `= i p`. The `spacemacs/evil-mc-paste-before` function only has two spaces, so they should probably match. Otherwise these lines will show up unexpectedly (for a git beginner), if the whole file is auto indented.

Additionally there's a small typo in the doc-string. 
"that" should be changed to "than" resulting in:
"Disable paste transient state if there is more than 1 cursor."
the same typo occurs in both functions.

Or should the typo be fixed in a separate PR?

File: layers/+spacemacs/spacemacs-evil/funcs.el [hunk 2]
```diff
@@ -84,16 +84,20 @@ Otherwise, revert to the default behavior (i.e. enable `evil-insert-state')."
...
+   (interactive "p")
+   (if (spacemacs//paste-transient-state-p)
+       (spacemacs/paste-transient-state/evil-paste-after)
+     (apply fn (first args) (or (second args) evil-this-register) (last args))))
 
-(defun spacemacs/evil-mc-paste-before (&optional count register)
+(advice-add 'evil-paste-after  :around  #'spacemacs/evil-mc-paste-after)
```
Comment (duianto) [inline layers/+spacemacs/spacemacs-evil/funcs.el]: There are two spaces before and after `:around`, is this just a style choice to make it more readable, or should there only be one? The same thing applies to the evil-paste-before advice below.
Searching through the .emacs.d folder found 5 other occurrences of `advice-add` ... `:around`, and they only have 1 space around them.

**Comment by duianto** [comment]
`C-n` and `C-p` now cycles through the kill ring after pasting with `p` or `P` 👍.

It does however not work after pasting with `[p` `evil-unimpaired/paste-above` or `]p` `evil-unimpaired/paste-below`, but they might not have worked before this issue occurred. They probably need to be fixed in a separate PR.

There's also a hiccup when cycling through the kill ring, it seems like when it reaches the end or beginning of the kill ring, then nothing happens until a second `C-n` or `C-p` is pressed. But I'm guessing that this also is unrelated to this issue and might even need to be fixed in the evil package.

**Comment by bmag** [comment]
> A more idiomatic way to create this functionality is to use elisp advice.

Using advices isn't idiomatic, it's more of a last resort. It might be merited here, I'm not sure.

The calls to `advice-add` should be in the :init block of `spacemacs-evil/init-evil-mc`, that way they don't apply when `evil-mc` is excluded.

Should use `funcall` with `count` and `register` instead of `apply` with `(first args)`, `(second args)` and `(last args)`. Function signatures should change accordingly from `(fn &rest args)` to `(fn &optional count register)`.

@duianto thanks for testing this

File: layers/+spacemacs/spacemacs-evil/funcs.el (hunk 1 above)
Comment (bmag) [inline layers/+spacemacs/spacemacs-evil/funcs.el]: > should the typo be fixed in a separate PR?

Either as part of this one or a separate PR, doesn't really matter.

**Comment by nochiel** [comment]
> It does however not work after pasting with [p evil-unimpaired/paste-above or ]p evil-unimpaired/paste-below, but they might not have worked before this issue occurred. They probably need to be fixed in a separate PR.

A number of Evil extensions also break `evil-paste-pop` in the same way that `spacemacs-evil` does.  Thus far, I count:
1. `evil-unimpaired`
2. `evil-iedit`
3. `evil-mc`

The correct fix would seem to be to write a new `evil-yank-handler` and/or setq `evil-last-paste` with sane values after the paste wrapper or yank handler runs (see, for example, what [evil-cleverparens](https://github.com/luxbock/evil-cleverparens/blob/82c920ba04accfd31fa292e11c454d5112b4fd51/evil-cleverparens.el#L419) does).  Currently it's too easy to break `evil-paste-pop` with the result that almost every extension that wraps it breaks it.  Alternatively, all these evil extensions could be PR'd with changes to their paste functions to `defadvice` them instead of `defun`.  But @bmag's insistence that advice is *not* idiomatic is given some credence by the elisp documentation which says that advice can lead to hard-to-diagnose bugs. (I happen to love advice and use it rather liberally in my own fiddling.)

> There's also a hiccup when cycling through the kill ring, it seems like when it reaches the end or beginning of the kill ring, then nothing happens until a second C-n or C-p is pressed. But I'm guessing that this also is unrelated to this issue and might even need to be fixed in the evil package.

Initially, I assumed that this was due to  [current-kill](https://www.gnu.org/software/emacs/manual/html_node/eintr/Code-for-current_002dkill.html#Code-for-current_002dkill) (used by `evil-paste-pop`):
>> Rotate the yanking point by N places, and then return that kill.      
>> If N is zero and `interprogram-paste-function' is set to a function that returns a string or a list of strings, and if that function doesn't return nil, then that string (or list) is added  to the front of the kill ring and the string (or first string in the list) is returned as the latest kill.

But a cursory test seems to indicate that this isn't the case (`yank-pop` seems to do the same thing). More testing required.

**Comment by nochiel** [comment]
> The calls to advice-add should be in the :init block of spacemacs-evil/init-evil-mc, that way they don't apply when evil-mc is excluded.

Done.

> Should use funcall with count and register instead of apply with (first args), (second args) and (last args). Function signatures should change accordingly from (fn &rest args) to (fn &optional count register).

Done

**Comment by nochiel** [comment]
Closing this and creating a new PR because there's a better simpler fix: ddd5499bc1aaeeae50fcedb7420229ccc42cdf9b
